### PR TITLE
New test for node draining (respecting PDB)

### PR DESCRIPTION
**Pull request by ingvagabund**
SSIA

### CODE REVIEW

File: test/e2e/provider_expectations.go [hunk 1]
```diff
@@ -298,3 +306,316 @@ func (tc *testConfig) ExpectNodeToBeDrainedBeforeDeletingMachine() error {
...
+									"cpu":    resource.MustParse("50m"),
+									"memory": resource.MustParse("50Mi"),
+								},
+							},
+						},
+					},
+					NodeSelector: nodeDrainLabels,
+					Tolerations: []corev1.Toleration{
+						{
+							Key:      "kubemark",
```
Comment (enxebre) [inline test/e2e/provider_expectations.go]: kubemark?

File: test/e2e/provider_expectations.go [hunk 2]
```diff
@@ -298,3 +306,316 @@ func (tc *testConfig) ExpectNodeToBeDrainedBeforeDeletingMachine() error {
 		return true, nil
 	})
 }
+
+var nodeDrainLabels = map[string]string{
+	workerRoleLabel:      "",
+	"node-draining-test": "",
+}
+
+func machineFromMachineset(machineset *mapiv1beta1.MachineSet) *mapiv1beta1.Machine {
```
Comment (enxebre) [inline test/e2e/provider_expectations.go]: I suspect you are doing this process because machineSet does not reconcile labels to machines today, so it'd be nice to add a comment/TODO so eventually we can add another test case for just scaling down a set

File: test/e2e/provider_expectations.go [hunk 3]
```diff
@@ -298,3 +306,316 @@ func (tc *testConfig) ExpectNodeToBeDrainedBeforeDeletingMachine() error {
...
+		glog.Infof("Waiting for RC ready replicas (%v/%v)", rcObj.Status.ReadyReplicas, rcObj.Status.Replicas)
+		if rcObj.Status.Replicas != rcObj.Status.ReadyReplicas {
+			return false, nil
+		}
+		return true, nil
+	}); err != nil {
+		return err
+	}
+
+	// All pods are distrubution evenly among all nodes so it's fine to drain
```
Comment (enxebre) [inline test/e2e/provider_expectations.go]: typo distrubution

File: test/e2e/provider_expectations.go (hunk 1 above)
Comment (ingvagabund) [reply test/e2e/provider_expectations.go]: Yes, so the same test can by used by both aws and kubemark. The toleration has no effect as long as node is not tainted.

File: test/e2e/provider_expectations.go (hunk 2 above)
Comment (ingvagabund) [reply test/e2e/provider_expectations.go]: It's done to get a machine template. This way I don't need to assume anything about machine provider config. I just take what is already provided by the cluster. In case of openshift installer I don't have a knowledge of which AMI I should use so I take one that is rendered by the installer.

File: test/e2e/provider_expectations.go (hunk 3 above)
Comment (ingvagabund) [reply test/e2e/provider_expectations.go]: Ups

File: test/e2e/provider_expectations.go [hunk 4]
```diff
@@ -298,3 +306,316 @@ func (tc *testConfig) ExpectNodeToBeDrainedBeforeDeletingMachine() error {
...
+		// one pod left on the node (that is being evicted).
+		if podCounter > 2 {
+			glog.Infof("Expecting at most 2 pods to be scheduled to drained node %v, got %v", machine.Status.NodeRef.Name, podCounter)
+			return false, nil
+		}
+
+		glog.Info("Expected result: all pods from the RC up to last one or two got scheduled to a different node while respecting PDB")
+		return true, nil
+	}); err != nil {
+		return err
```
Comment (enxebre) [inline test/e2e/provider_expectations.go]: we are we validating the machine object and the backed node does not exist anymore?

Comment (ingvagabund) [reply test/e2e/provider_expectations.go]: if the machine object does not get removed, the loop above fails due to
```go
if err := tc.client.Get(context.TODO(), key, &machine); err != nil {
	glog.Errorf("error querying api machine %q object: %v, retrying...", machine1.Name, err)
	return false, nil
}
```

The backed up node is actually removed later (by cloud controller manager once it goes NotReady). It can happen in 1 minute, in 2 minutes, later, sooner? So, we can't verify that.

Comment (enxebre) [reply test/e2e/provider_expectations.go]: yes we can, and we should. We are actually validating that in other test cases

Comment (ingvagabund) [reply test/e2e/provider_expectations.go]: We can do it though this test is about verifying a node is drained before machine object is deleted. It does not test a node is deleted. Since testing "node is deleted after linked machine is delete" is another test case.

**Comment by bison** [comment]
I find this a little hard to follow, but I think it looks okay. One thing I don't see us checking is that the node is marked as unscheduable. The drain should mark the node unscheduable then create evictions when it can. Should we check for that directly rather than just watching pod counts?

File: test/e2e/provider_expectations.go (hunk 4 above)
Comment (enxebre) [reply test/e2e/provider_expectations.go]: This should test e2e features from a product end user pov. This test is pretty much validating "Pod disruption budget" and eviction which is only a part of the user expected e2e feature. Then it is assuming that the pods were moved because the machine was deleted (but that could be untrue. Correct draining could be a result of buggy code on a controller telling to drain but then failing to delete the machine and as a result the node).
We want to validate in code explicitly exactly the end user story: When I delete a machine the node is drained (covered by the test), then machine is deleted (not covered), then node is delete  (not covered)

**Comment by ingvagabund** [comment]
> I find this a little hard to follow, but I think it looks okay. One thing I don't see us checking is that the node is marked as unscheduable. The drain should mark the node unscheduable then create evictions when it can. Should we check for that directly rather than just watching pod counts?

We might check if a node is `unscheduable` in addition to what we have now. Yet, it does not guarantee pods already scheduled on the node to be re-scheduled some place else. Draining operation makes sure all pods are properly evicted before a node is removed. So the main goal of the test is to verify all relevant pods (excluding daemon set pods) are removed from the drained node while still making sure RC has at most one pod unready. So we need to check the pod count as well.

**Comment by bison** [comment]
> We might check if a node is `unscheduable` in addition to what we have now. Yet, it does not guarantee pods already scheduled on the node to be re-scheduled some place else. Draining operation makes sure all pods are properly evicted before a node is removed. So the main goal of the test is to verify all relevant pods (excluding daemon set pods) are removed from the drained node while still making sure RC has at most one pod unready. So we need to check the pod count as well.

Right, I get that, but I guess that's kind of my point. That sounds like testing the eviction or scheduling code in a way, not ours. I think we mainly care that the drain process was initiated and that we don't remove the machine prematurely. It's unlikely, but *if* something else caused the pods to be rescheduled, this could pass. But it's kind of a weird situation because "draining" isn't really a concept in the API, it's just a series of steps some tools take.

Anyway, not saying this is wrong, just that anything we can do to check more directly that `Drain()` was called is nice.

**Comment by enxebre** [comment]
/approve
Agree with @bison that we want approach drain abstraction

**Comment by ingvagabund** [comment]
https://github.com/openshift/cluster-api-provider-aws/pull/163 needs to be merged before this gets
